Docs: Getting started (#887)
zio/zio-http: docs/website/docs/v1.x/getting-started.md

# Getting Started

> ZIO HTTP is a powerful library that is used to build highly performant HTTP-based services and clients using functional scala and ZIO and uses Netty as its core.
> The first step when using ZIO HTTP is creating an HTTP app. 
> ZIO HTTP has some powerful functional domains which help in creating, modifying, composing apps easily.
> Let's start with the HTTP domain.
> 

Reviewer: ```suggestion
**ZIO HTTP** is a powerful library that is used to build highly performant HTTP-based services and clients using functional scala and ZIO and uses [Netty](https://netty.io/) as its core.
ZIO HTTP has powerful functional domains which help in creating, modifying, composing apps easily. Let's start with the HTTP domain.
The first step when using ZIO HTTP is creating an HTTP app. 
```

## Http

HTTP is a domain that models Http apps using ZIO and works over any request and response types. 

> Http Domain provides a lot of ways to create HTTP apps, for example `Http.text`, `Http.html`, `Http.fromFile`, `Http.fromData`, `Http.fromStream`, `Http.fromEffect`. 

Reviewer: ```suggestion
`Http` is a domain that models HTTP apps using ZIO and works over any request and response types. `Http` Domain provides different constructors to create HTTP apps, `Http.text`, `Http.html`, `Http.fromFile`, `Http.fromData`, `Http.fromStream`, `Http.fromEffect`. 
```

### Creating a "_Hello World_" app

Creating an HTTP app using ZIO Http is as simple as this:

> 

Reviewer: ```suggestion
Creating an HTTP app using ZIO Http is as simple as given below, this app will always respond with "Hello World!"

```

```scala
import zhttp.http._

val app = Http.text("Hello World!")
```

> In the above snippet, for any HTTP request, the response is always `"Hello World!"`.

Reviewer: ```suggestion
```

> An application can be made using any of the available operators on `zhttp.Http`.

Reviewer: ```suggestion
An app can be made using any of the available constructors on `zhttp.Http`.
```

### Routing

 For handling routes, Http Domain has a `collect` method that, accepts different requests and produces responses. Pattern matching on the route is supported by the framework
The example below shows how to create routes:

```scala,
import zhttp.http._

val app = Http.collect[Request] {
  case Method.GET -> !! / "fruits" / "a"  => Response.text("Apple")
  case Method.GET -> !! / "fruits" / "b"  => Response.text("Banana")
}
```
You can create typed routes as well. The below example shows how to accept count as `Int` only.
 ```scala,
 import zhttp.http._
 
 val app = Http.collect[Request] {
   case Method.GET -> !! / "Apple" / int(count)  => Response.text(s"Apple: $count")
 }
 ```

### Composition

> HTTP app can be composed using the `++` operator. The way it works is if none of the routes matches in `a` or there is any error in `a`, the control passes on to the `b` app.

Reviewer: ```suggestion
HTTP app can be composed using the `++` operator. The way it works is if none of the routes matches in `a` or there is an error `a`, the control is passed to the `b` app.
```

### ZIO Integration

For creating effectful apps, you can use `collectZIO` and wrap `Response` using `wrapZIO` to produce ZIO effect value.

```scala
val app = Http.collectZIO[Request] {
  case Method.GET -> !! / "hello" => Response.text("Hello World").wrapZIO
}
```

### Accessing the Request

> To access request in the response, use `@` as it binds a matched pattern to a variable and can be used while creating a response.  

Reviewer: ```suggestion
To access the request use `@` as it binds a matched pattern to a variable and can be used while creating a response.
```

      testM("should be ok") {
        val app = Http.ok
        val req = Request()
        assertM(app(req))(equalTo(Response.ok))
      }
    )
}
```

> When we call `app` with `request` it calls apply the method of HTTP via `zhttp.test` package

Reviewer: ```suggestion
When we call the `app` with the `request` it calls the apply method of `Http` via `zhttp.test` package
```

## Socket

> `Socket` is another functional domain in ZIO HTTP. It provides operators to create socket apps. 
> A socket app is an app that handles WebSocket connections.

Reviewer: ```suggestion
`Socket` is functional domain in ZIO HTTP. It provides constructors to create socket apps. 
A socket app is an app that handles WebSocket connections.
```

## Server

As we have seen how to create HTTP apps, the only thing left is to run an  HTTP server and serve requests.
ZIO HTTP provides a way to set configurations for your server. The server can be configured according to the leak detection level, request size, address etc. 

### Starting an HTTP App

> To launch our app, we need to start the server on some port. The below example shows a simple HTTP app that responds with empty content and a `200` status code, deployed on port `8090` using `Server.start`.

Reviewer: ```suggestion
To launch our app, we need to start the server on a port. The below example shows a simple HTTP app that responds with empty content and a `200` status code, deployed on port `8090` using `Server.start`.
```
